### PR TITLE
Oracle adapter initial commit

### PR DESCRIPTION
**Pull request by taoenkh**
# Oracle Adapter #

## Overview
Develop another adapter for Cloudberry to support Oracle as a backend database.

## Note
Since `OracleDB` does not support `if exists` syntax, writing `PL/SQL` will be required. The `PL/SQL` requires checking the table space, so the `Oracle` user must create an user named `BERRY` to use the `OracleAdapter`.

# Summary
### Major differences from `MySQL` adapter
Changes | Explanation
------------ | -------------
Convert `if exists` to `pl/sql` | `Oracle` 's syntax does not support keyword `if exists`, creating `berry.meta`, dropping table, and creating view's `if exists` should be replaced to `pl/sql`. (Raw Data / Meta Data)
`limit` | In `Oracle` `fetch first n rows` is equivalent to `limit`(Raw Data/Meta Data)
`replace into` | `replace in to` is used when creating a view in the database. `merge into` can be used in `Oracle instead` (Raw Data / Meta Data)
`group by` | In `MySQL` the select clause executes first, but in `Oracle` where clause executes first. Changes in the function of parsing group by. (Raw Data)
Extracting time units | In `MySQL` there are functions to extract a certain time field, and the usage in `Oracle` is different. (Raw Data)
Full text search |  Different syntax (Raw Data)
Adding quotations | Currently, names such as `berry.meta`, `stats.creatTime` are all using quotations to correspond to `AsterixDB`(Same in `MySQL`) (Meta Data / Raw Data) 
Geo DataType | In `AsterixDB` the built-in data type `Point` could be queried by directly using `select t.coordinate`, but `Oracle`'s `SDO_GEOMETRY` type requires different syntax 

## Functions overrode
- `parseTimeRelation`
- `parseTextRelation`
- `parseDrop`
- `parseCreate`
- `parseSelect`
- `parseGroupByFunc`
- `parseGroupBy`
- `parseUpssertMeta`
- `fieldType2SQLType`
- `postGeneralQuery` in `OracleConn`

## Quick Guide for using this Oracle Adapter on MAC OS X
 - Since Oracle DB does not support latest MAC OS X, we install Oracle 12c on a Linux machine, and run Cloudberry and Twittermap on a MAC machine.

### 1 Deploy Oracle DB
_(*NOTE*: This is the only step done on a Linux machine.)_
#### 1.1 Install Oracle DB
Following link is a sample guidance for installing Oracle DB 12c on CentOS 7:
https://www.tecmint.com/install-oracle-database-12c-on-centos-7/
If you are using other versioned Linux/Windows systems, please find relevant resources online.
#### 1.2 Create user `berry` in Oracle
```plsql
create user berry identified by orcl;
grant connect, resource, dba to berry;
```

#### 2.1 Install python 3.7 and `cx_Oracle` 
```sh
brew install python
pip3 install cx_Oracle --upgrade --user
```
(reference: https://cx-oracle.readthedocs.io/en/latest/installation.html#installing-cx-oracle-on-macos)

#### 2.2 Ingest Tweets into Oracle DB

```sh
cd cloudberry/examples/twittermap
# If you have different username/password or hostname or SID or your oracle configuration, you'll need to modify the link [con = cx_Oracle.connect('berry/orcl@localhost:1521/orcl')] inside those two scripts: ingestOracledata.py, ingestpopulationOracle.py
./script/OracleTwittermapIngestion.sh
```

### 3 Configure Cloudberry and Run

Modify `cloudberry/neo/conf/applicaiton.conf`:

```sh
#asterixdb.lang = SQLPP # comment this line
asterixdb.lang = oracle # uncomment this line
oracledb.url = "jdbc:oracle:thin:berry/orcl@localhost:1521:orcl" # uncomment this line with url of your oracle
```

Done!

Now you can start Cloudberry and Twittermap as before.

### CODE REVIEW

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+      }
+    }
+    val fields = schema.fieldMap.values.filter(f => f.dataType != DataType.Hierarchy && f != AllField).map {
+      f => mkNestDDL(f.name, fieldType2SQLType(f) + (if (f.isOptional) " default null" else " not null"))
+    }
+    s"""
+       |declare
+       |    result1 number(8);
+       |begin
+       |    select count(*) into result1 from dba_tables where owner = 'BERRY' and table_name = '${name}';
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Same concern here for `dba_tables`

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+
+        val producedExprs = mutable.LinkedHashMap.newBuilder[String, FieldExpr]
+
+        val orderStrs = select.orderOn.zip(select.order).map {
+          case (orderOn, order) =>
+            val expr = exprMap(orderOn.name).defExpr
+            val orderStr = if (order == SortOrder.DSC) "desc" else ""
+            s"${expr} $orderStr"
+        }
+        val orderStr =
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Better to have a newline here.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+          if (orderStrs.nonEmpty) {
+            orderStrs.mkString("order by ", ",", "")
+          } else {
+            s""
+          }
+        val limitStr = s"fetch first ${select.limit} rows only"
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: new line.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+        appendIfNotEmpty(queryBuilder, orderStr)
+        if (select.limit != 0) {
+          appendIfNotEmpty(queryBuilder, limitStr)
+        }
+        if (select.fields.isEmpty || query.hasUnnest || query.hasGroup) {
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: new line

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+          producedExprs ++= exprMap
+        } else {
+          select.fields.foreach {
+            field => {
+              if (field.dataType != DataType.Point) {
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Same concern, if `ojdbc` supports `SDO_GEOMETRY` type, we don't need to convert them into `VARCHAR2`.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+                producedExprs += field.name -> exprMap(field.name)
+              }
+              else{
+                val fieldExpr = exprMap(field.name)
+                producedExprs += field.name -> FieldExpr(field.name,s"""concat(concat(concat('POINT(',${fieldExpr.refExpr}.sdo_point.x),concat(', ',${fieldExpr.refExpr}.sdo_point.y)),')')""")
+              }
+            }
+          }
+        }
+        val newExprMap = producedExprs.result().toMap
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: new line

File: examples/twittermap/web/conf/application.conf [hunk 7]
```diff
@@ -91,7 +91,7 @@ sentimentUDF = "twitter.`snlp#getSentimentScore`(text)"
 # Customize TwitterMap for a specific domain
 removeSearchBar = false
 predefinedKeywords = ["Diabetes", "Sleep Disorder"]
-startDate = "2017-01-24T00:00:00.000"
+startDate = "2016-01-24T00:00:00.000"
```
Comment (baiqiushi) [inline examples/twittermap/web/conf/application.conf]: Let's do not submit irrelevant changes.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+      }
+      else {
+        dimensionStrs = dimensionStrs:+"s.\"" + d.name + "\""
+      }
+    }
+    var conflictsting = ""
+    for (pkname<- primaryKeystr){
+      conflictsting += "d.\""+pkname+"\" = s.\"" + pkname+"\" and "
+    }
+    val insert = s"""
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Is this whole construction of `insert` statement the same as previous `parseAppend` function? If so, can we reuse them by abstracting the code to a new function or using one of them.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    * @param field
+    */
+  protected def fieldType2SQLType(field: Field): String = {
+    field.dataType match {
+      case DataType.Number => "NUMBER"
+      case DataType.Time => "TIMESTAMP"
+      case DataType.Point => "SDO_GEOMETRY"
+      case DataType.Boolean => "NUMBER(1)"
+      case DataType.String => "VARCHAR2(255)"
+      case DataType.Text => "VARCHAR(1000)"
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: `Text` in AsterixDB can NOT be matched a fixed length `VARCHAR` in Oracle, which will cause trouble if the length of values exceed the 1000 length which is not rare. We may need to consider `clob` in Oracle.

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+    }
+  }
+
+  //TODO: unnest
+  protected def parseUnnest(unnest: Seq[UnnestStatement],
+                            exprMap: Map[String, FieldExpr], queryBuilder: StringBuilder): ParsedResult = {
+    //return the empty result & exprMap for next step's process.
+    ParsedResult((new ListBuffer[String]), exprMap)
+  }
+  override protected def timeUnitFuncMap(unit: TimeUnit.Value): String = unit match {
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Isn't it the same as the definition in `SQLGenerator`? If so why do we need to override it?

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+    case TimeUnit.Second => "second"
+    case TimeUnit.Minute => "minute"
+    case TimeUnit.Hour => "hour"
+    case TimeUnit.Day => "date"
+    case TimeUnit.Month => "month"
+    case TimeUnit.Year => "year"
+    case _ => throw new QueryParsingException(s"No implementation is provided for timeunit function ${unit.toString}")
+  }
+  protected def parseGroupByFunc(groupBy: ByStatement, fieldExpr: String): String = {
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: new line

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+    groupBy.funcOpt match {
+      case Some(func) =>
+        func match {
+          case interval: Interval => {
+            interval.unit match {
+              case TimeUnit.Day=>s"to_char(cast($fieldExpr as ${timeUnitFuncMap(interval.unit)}),'YYYY-MM-DD')"
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Why only `Day` is different from other time unit types?

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+              case _ => s"extract(${timeUnitFuncMap(interval.unit)} from $fieldExpr)"
+            }
+          }
+          case level: Level =>
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: What does this `Level` do?

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+
+
+  override protected def parseDrop(query: DropView, schemaMap: Map[String, AbstractSchema]): String = {
+
+
+    s"""
+       |declare
+       |     result1 number(8);
+       |begin
+       |   select count(*)into result1 from dba_tables where owner = 'BERRY' and table_name = '${query.dataset}';
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Same concern for `dba_tables`

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+       |
+       | """.stripMargin
+
+
+  }
+
+
+  /**
+    * Process SDO_POINT_TYPE of ORACLE
+    * : return POINT field as text to avoid messy code. https://dev.mysql.com/doc/refman/5.7/en/gis-format-conversion-functions.html
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/OracleGenerator.scala]: Let's remove those comments specific to `mysql`.

File: cloudberry/zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/TwitterDataStoreWithHashTag.scala [hunk 16]
```diff
@@ -6,44 +6,44 @@ object TwitterDataStoreWithHashTag {
...
-                                      NumberField("in_reply_to_user"),
-                                      NumberField("favorite_count"),
-                                      NumberField("retweet_count"),
-                                      BagField("hashtags", DataType.String, true),
-                                      BagField("user_mentions", DataType.Number, true)
-                                    ),
-                                    Seq(NumberField("id")),
-                                    TimeField(TimeFieldName))
-}
+    Seq(
```
Comment (baiqiushi) [inline cloudberry/zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/TwitterDataStoreWithHashTag.scala]: Are these changes necessary?

File: examples/twittermap/script/ingestOracledata.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
+# -*- coding: utf-8 -*-
+import os
+os.environ['NLS_LANG'] = 'SIMPLIFIED CHINESE_CHINA.UTF8'
```
Comment (waans11) [inline examples/twittermap/script/ingestOracledata.py]: Why is this setting required?

File: examples/twittermap/script/ingestpopulationOracle.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+import os
+os.environ['NLS_LANG'] = 'SIMPLIFIED CHINESE_CHINA.UTF8'
```
Comment (waans11) [inline examples/twittermap/script/ingestpopulationOracle.py]: ditto

File: cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLGenerator.scala [hunk 19]
```diff
@@ -171,6 +171,8 @@ abstract class SQLGenerator extends IQLGenerator {
 
     val resultAfterGlobalAggr = parseGlobalAggr(query.globalAggr, resultAfterSelect.exprMap, queryBuilder)
 
+    println(queryBuilder.toString)
```
Comment (baiqiushi) [inline cloudberry/zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLGenerator.scala]: Get rid of this output for debugging.

File: cloudberry/project/dependencies.scala [hunk 20]
```diff
@@ -27,6 +27,8 @@ object Dependencies {
     "joda-time" % "joda-time" % "2.9.3",
     "com.typesafe.akka" %% "akka-actor" % "2.4.4",
     "mysql" % "mysql-connector-java" % "5.1.24",
+    "com.oracle" % "ojdbc8"%"12.1.0.2",
```
Comment (baiqiushi) [inline cloudberry/project/dependencies.scala]: We'll move this file's changes to our Readme.txt file, because without the `maven` configuration thing, normal users who don't want to use Oracle can NOT build our project anymore.
